roundabout/functions/redirect.js

```diff
@@ -33,14 +34,31 @@ export async function redirectCarGet(request) {
       statusCode: 400
     }
   }
-
-  const signer = getSigner(s3Client, BUCKET_NAME)
-  const key = `${cid}/${cid}.car`
-  const signedUrl = await signer.getUrl(key, {
-    expiresIn
-  })
   
-  return toLambdaResponse(signedUrl)
+  const signer = getSigner(s3Client, BUCKET_NAME)
+
+  if (asPieceCid(cid) !== undefined) {
+    const cars = await findEquivalentCarCids(cid)
+    for (const carCid of cars) {
+      const key = `${carCid}/${carCid}.car`
+      // getUrl returns undefined if we don't have that car, so keep trying till we find a good one.
+      const signedUrl = await signer.getUrl(key, { expiresIn }) 
+      if (signedUrl) {
+        return toLambdaResponse(signedUrl)
+      }
+    }
```

Comment:
Nit: I personally prefer more linear logic so I don't have to maintain stack in my head, e.g. in this case return here would make it easier to follow as I don't have to consider next if block.

Better yet would be to refactor logic paths into separate functions e.g.

```js
const resolvePiece = async (cid, { signer, expiresIn }) => {
  const piece = asPieceCid(cid)
  if (piece) {
    const cars = await findEquivalentCarCids(piece)
    // Attempt to find a CAR that we can resolve
    for (const car of cars) {
      const url = await signer.getUrl(`${car}/${car}.car`, { expiresIn })
      if (url) {
        return url
      }
    }
    // We don't have a CAR for this piece CID
    return null
  } else {
    return null
  }
}

const resolveContentArcive = async (cid, {signer, expiresIn}) => {
   // ...
}

export async function redirectCarGet(request) {
   // ...
   return await resolvePiece(cid, {signer, expiresIn}) ??
          await resolveContentArcive(cid, { signer, expiresIn }) ??
          await notFound(cid)
}
```

Reply:
Nit: It also might be a good idea to return 404 with some error message explaining that we could not find a CAR for the piece. That way e.g. if user provides PieceV1 CID they'll be able to tell that we simply do not support those versions as opposed to us not having a content for them.

Reply:
I prefer to have less state to think about too. I've refactored this to be clearer.